purchase_discount/models/product_supplierinfo.py

```diff
@@ -32,7 +32,7 @@ def create(self, vals_list):
         """ Insert discount (or others) from context from purchase.order's
         _add_supplier_to_product method """
         for vals in vals_list:
-            product_tmpl_id = vals['product_tmpl_id']
+            product_tmpl_id = vals.get('product_tmpl_id', False)
```

Comment:
I don't understand in wich case the value is not set. 

I think that if not defined, Odoo will not work correctly, as the link between the product.template and the product.supplierinfo is lost. don't you think ? 

https://github.com/odoo/odoo/blob/12.0/addons/product/models/product_template.py#L114

Reply:
@legalsylvain 
If there is a custom module that creates a product.supplierinfo then link it to the product the vals on create function can be without product_tmpl_id